spharpy/classes/audio.py

```diff
@@ -1,12 +1,328 @@
+    normalization : str
+        Normalization convention, either ``'N3D'``, ``'NM'``,
+        ``'maxN'``, ``'SN3D'`` or ``'SNM'``. (maxN is only supported up
+        to 3rd order)
+    channel_convention : str
+        Channel ordering convention, either ``'ACN'`` or ``'FuMa'``.
+        (FuMa is only supported up to 3rd order)
+
+    Returns
+    -------
+    numpy.ndarray
+        Spherical harmonic data according to the desired definition.
+    """
+
+    data = renormalize(
+        data, "ACN", "N3D", normalization, axis=-2)
+
+    data = change_channel_convention(
+        data, "ACN", channel_convention, axis=-2)
+
+    return data
+
+
+class _SphericalHarmonicAudio(_Audio, _SphericalHarmonicBase, ABC):
+    """
+    Base class for spherical harmonics audio objects.
+
+    This class extends the pyfar Audio class with all methods and
+    properties required for spherical harmonics data and are common to the
+    three sub-classes :py:func:`SphericalHarmonicsTimeData`,
+    :py:func:`SphericalHarmonicsFrequencyData`, and
+    :py:func:`SphericalHarmonicsSignal`.
+
+    Objects of this class contain spherical harmonics coefficients which are
+    directly convertible between channel conventions ACN and FUMA, as
+    well as the normalizations N3D, SN3D, or MaxN. The definition of
+    the spherical harmonics basis functions is based on the scipy convention
+    which includes the Condon-Shortley phase.
+
+
+    Parameters
+    ----------
+    basis_type : str
+        Type of spherical harmonic basis, either ``'complex'`` or
+        ``'real'``.
+    normalization : str
+        Normalization convention, either ``'N3D'``, ``'NM'``,
+        ``'maxN'``, ``'SN3D'`` or ``'SNM'``. (maxN is only supported up
+        to 3rd order)
+    channel_convention : str
+        Channel ordering convention, either ``'ACN'`` or ``'FuMa'``.
+        (FuMa is only supported up to 3rd order)
+    condon_shortley : bool
+        Flag to indicate if the Condon-Shortley phase term is included
+        (``True``) or not (``False``).
+    domain : ``'time'``, ``'freq'``, optional
+        Domain of data. The default is ``'time'``
+    comment : str
+        A comment related to `data`. The default is ``None``.
+
+    """
+
+    def __init__(self, basis_type, normalization, channel_convention,
+                 condon_shortley):
+
+        _SphericalHarmonicBase.__init__(
+            self,
+            basis_type,
+            channel_convention,
+            normalization,
+            condon_shortley)
+
+    @property
+    def n_max(self):
+        """Get or set the spherical harmonic order."""
+        return int(np.sqrt(self.cshape[-1])-1)
+
+    @property
+    def basis_type(self):
+        """Get or set the type of spherical harmonic basis."""
+        return _SphericalHarmonicBase.basis_type.fget(self)
+
+    @basis_type.setter
+    def basis_type(self, value):
+        """Get or set the type of spherical harmonic basis."""
```

Comment:
Its possible to overwrite only the setter, which reduced duplicate Code:
```suggestion
    @_SphericalHarmonicBase.basis_type.setter
    def basis_type(self, value):
        """Get or set the type of spherical harmonic basis."""

        # Make sure that the basis type can only be set during initialization.
        # Changing it afterwards requires implementing the conversion between
        # real and complex-valued coefficients, which is possible but not yet
        # implemented.
```